ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/ScriptHandlers/Sandboxed/BraveLeoScriptHandler.swift

```diff
@@ -0,0 +1,93 @@
+import os.log
+import AIChat
+
+class BraveLeoScriptHandler: NSObject, TabContentScript {
+  fileprivate weak var tab: Tab?
+
+  init(tab: Tab) {
+    self.tab = tab
+    super.init()
+  }
+  
+  static let getMainArticle = "getMainArticle\(uniqueID)"
+  static let getPDFDocument = "getPDFDocument\(uniqueID)"
+
+  static let scriptName = "BraveLeoScript"
+  static let scriptId = UUID().uuidString
+  static let messageHandlerName = "\(scriptName)_\(messageUUID)"
+  static let scriptSandbox: WKContentWorld = .defaultClient
+  static let userScript: WKUserScript? = {
+    guard var script = loadUserScript(named: scriptName) else {
+      return nil
+    }
+    
+    return WKUserScript(source: secureScript(handlerNamesMap: ["$<message_handler>": messageHandlerName,
+                                                               "$<getMainArticle>": getMainArticle,
+                                                               "$<getPDFDocument>": getPDFDocument],
+                                             securityToken: scriptId,
+                                             script: script),
+                        injectionTime: .atDocumentEnd,
+                        forMainFrameOnly: true,
+                        in: scriptSandbox)
+  }()
+
+  func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage, replyHandler: (Any?, String?) -> Void) {
+    defer { replyHandler(nil, nil) }
+    
+    if !verifyMessage(message: message) {
+      assertionFailure("Missing required security token.")
+      return
+    }
+  }
+}
+
+extension BraveLeoScriptHandler: AIChatJavascript {
+  
+  @MainActor
+  static func getPageContentType(webView: WKWebView) async -> String? {
+    return try? await webView.evaluateSafeJavaScriptThrowing(functionName: "document.contentType", contentWorld: Self.scriptSandbox, escapeArgs: false, asFunction: false) as? String
+  }
+  
+  @MainActor
+  static func getMainArticle(webView: WKWebView) async -> String? {
+    do {
+      let articleText = try await webView.evaluateSafeJavaScriptThrowing(functionName: "window.__firefox__.\(getMainArticle)",
+                                                                         args: [Self.scriptId],
+                                                                         contentWorld: Self.scriptSandbox,
+                                                                         asFunction: true) as? String
+      return articleText
+    } catch {
+      Logger.module.error("Error Retrieving Main Article From Page: \(error.localizedDescription)")
+      return nil
+    }
+  }
+  
+  @MainActor
+  static func getPDFDocument(webView: WKWebView) async -> String? {
+    // Pages containing PDF cannot contain injected Javascript
+    // So we must execute an inline script
+    return try? await webView.callAsyncJavaScript(
+    """
+    const buffer = await window.fetch(window.location.href, {
+      method: 'GET',
```

Comment:
The PDF itself can be served after a POST request. Let's say you fill some data on a website and press submit, it will send a POST request and return a PDF in response. When you refetch it like this with a GET request it may not work at all, since the webserver expects a POST request instead. Here is a Python example, you need to install
```
pip install Flask
pip install fpdf
```
to run it

```python
from fpdf import FPDF
from flask import Flask, request, render_template_string, redirect, url_for, send_file
from io import BytesIO

app = Flask(__name__)

@app.route('/')
def index():
    return render_template_string('''
    <!DOCTYPE html>
    <html>
    <head>
        <title>Generate PDF</title>
    </head>
    <body>
        <form action="{{ url_for('generate_pdf') }}" method="post">
            <button type="submit">Generate PDF</button>
        </form>
    </body>
    </html>
    ''')

@app.route('/generate_pdf', methods=['POST'])
def generate_pdf():
    # create a PDF in memory
    pdf = FPDF()

    # add a page
    pdf.add_page()
    pdf.set_font('Arial', 'B', 14)
    pdf.cell(40, 10, "Hello World!")
    byte_string = pdf.output(dest="S")
    b = BytesIO(byte_string.encode('latin-1'))
    # send the PDF as a response
    return send_file(b, 
mimetype='application/pdf', 
download_name='my_pdf.pdf')

if __name__ == '__main__':
    app.run(debug=True)
```

Go to http://127.0.0.1:5000, press `Generate PDF` and then try to as Leo to summarize it. I expect it won't be able to refetch the PDF since the server will return Method Not Allowed on GET requests

Reply:
I mean it's an edge case, but it's not uncommon to server PDFs after a POST request

Reply:
Addressed here: https://github.com/brave/brave-core/pull/22200/commits/c8212db2218d089aaf8612ca4ab7af03b01daa11

We are too limited to handle `POST` requests atm. It will always hit the server if I try. If the server has `no-cache` then it will hit every time and it could take long.

I found a way to steal the PDF from WebKit out of process :)
Which means the above Javascript isn't needed, but also I want to keep it as a fallback in case WebKit's Private-API changes.

At least this way, now the Javascript never gets hit, and we don't have to worry about cache-miss or anything as it takes it directly from WebKit.

Reply:
A nice find 🙂